axum/src/routing/mod.rs

```diff
@@ -111,22 +113,47 @@ where
     /// all requests.
     pub fn new() -> Self {
         Self {
-            path_router: Default::default(),
-            fallback_router: PathRouter::new_fallback(),
-            default_fallback: true,
-            catch_all_fallback: Fallback::Default(Route::new(NotFound)),
+            inner: Arc::new(RouterInner {
+                path_router: Default::default(),
+                fallback_router: PathRouter::new_fallback(),
+                default_fallback: true,
+                catch_all_fallback: Fallback::Default(Route::new(NotFound)),
+            }),
+        }
+    }
+
+    fn update_router<F, S2>(self, f: F) -> Router<S2>
```

Comment:
I would call this `map_inner`. Maybe we should also have a `tap_inner[_mut]` for an `FnOnce(&mut RouterInner<S>)` that's implemented in terms of this fn, would simplify a few call sites.